service/src/main/java/io/camunda/service/license/CamundaLicense.java

```diff
@@ -7,49 +7,76 @@
  */
 package io.camunda.service.license;
 
+import io.camunda.zeebe.util.VisibleForTesting;
 import org.camunda.bpm.licensecheck.InvalidLicenseException;
 import org.camunda.bpm.licensecheck.LicenseKey;
 import org.camunda.bpm.licensecheck.LicenseKeyImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class CamundaLicense {
 
-  public String license;
-  public boolean isValid;
+  private static final Logger LOGGER = LoggerFactory.getLogger(CamundaLicense.class);
+  private boolean isValid;
+  private boolean isSelfManaged;
   private volatile boolean isInitialized;
 
-  public CamundaLicense(final String license) {
-    this.license = license;
-  }
-
-  // for testing
   public CamundaLicense() {}
 
   public boolean isValid() {
-    initializeLicenseKey();
     return isValid;
   }
 
-  protected void initializeLicenseKey() {
+  public boolean isSelfManaged() {
+    return isSelfManaged;
+  }
+
+  public void initializeWithLicense(final String license) {
     if (!isInitialized) {
       synchronized (this) {
         if (!isInitialized) {
+
           isValid = determineLicenseValidity(license);
+          isSelfManaged = determineIfLicenseEnvModeIsSelfManaged(license);
+
           isInitialized = true;
         }
       }
     }
   }
 
-  protected boolean determineLicenseValidity(final String licenseStr) {
+  /**
+   * SaaS mode is determined through the properties of the passed in license.
+   *
+   * <p>Self-managed mode is any other possibility. (ex, blank license, prop missing, etc)
+   */
+  private boolean determineIfLicenseEnvModeIsSelfManaged(final String licenseStr) {
+    try {
+      final LicenseKey licenseKey = getLicenseKey(licenseStr);
+      return licenseKey.getProperties().entrySet().stream()
+          .noneMatch(x -> x.getKey().equals("environmentMode") && x.getValue().equals("saas"));
+    } catch (final InvalidLicenseException e) {
+      return true;
+    } catch (final Exception e) {
+      LOGGER.error("Unexpected error when determining environment mode of license", e);
+      return true;
```

Comment:
💭 Wondering whether we maybe should explode/fail in this case instead of continuing? Is it expected that the licenseStr is invalid? 🤔

Reply:
When encountering an invalid license, we don't blow up the startup process but rather continue on and mark the license as invalid. I'd want to treat an unexpected error the same, there is something wrong with the license, and we should consider it invalid